runtime/interpreter/statictype.go

```diff
@@ -477,6 +511,25 @@ outer:
 	return t.Type.Equal(otherRestrictedType.Type)
 }
 
+func (t *RestrictedStaticType) ID() TypeID {
+	if t.typeID == "" {
+		var restrictionStrings []string
+		restrictionCount := len(t.Restrictions)
+		if restrictionCount > 0 {
+			restrictionStrings = make([]string, 0, restrictionCount)
+			for _, restriction := range t.Restrictions {
+				restrictionStrings = append(restrictionStrings, string(restriction.ID()))
+			}
+		}
+		var typeString string
+		if t.Type != nil {
+			typeString = string(t.Type.ID())
+		}
+		t.typeID = TypeID(sema.FormatRestrictedTypeID(typeString, restrictionStrings))
+	}
+	return t.typeID
```

Comment:
The type ID for restricted types depends on the order of the restrictions. I guess that is OK, because this simply replaces sema.Type.ID(), and that too relies on the order?

Reply:
Right, this is a current issue for restricted types, see https://github.com/onflow/cadence/issues/2681.
The code here keeps the current order dependency as it is currently, i.e. sema's type ID encoding for restricted types is also based on declaration order / performs no sorting.